File: frontend/src/types/monitoring.ts
export interface TemporalSignal {
  metric_name: string;
  record_count: number;
  latest_year: number | null;
  latest_value: number;
  yoy_change_pct: number;
  yoy_change_absolute: number;
  rolling_3yr_mean: number;
  rolling_3yr_std: number;
  rolling_3yr_zscore: number;
  rolling_5yr_mean: number;
  rolling_5yr_std: number;
  rolling_5yr_zscore: number;
  rolling_8yr_mean: number;
  rolling_8yr_std: number;
  trend_slope: number;
  acceleration: number;
  volatility_cv: number;
  historical_mean: number;
  deviation_from_historical: number;
  trajectory: Array<{
    year: number;
    value: number;
    historical_mean: number;
    rolling_3yr: number;
    yoy_change_pct: number;
  }>;
  location?: string;
  state?: string;
  district?: string;
}

export interface EarlyWarning {
  signal_id: string;
  signal_type: string;
  state: string;
  district: string;
  year: number;
  severity: 'INFO' | 'WATCH' | 'ELEVATED' | 'HIGH' | 'CRITICAL';
  trigger_value: number;
  threshold: number;
  unit: string;
  evidence: string[];
  recommended_action: string;
  is_active: boolean;
}

export interface AlertValidationContext {
  r2: number;
  mae: number;
  drift_status: string;
  data_quality_score: number;
}

export interface Alert {
  alert_id: string;
  location: string;
  state: string;
  district: string;
  year: number;
  severity: 'INFO' | 'WATCH' | 'ELEVATED' | 'HIGH' | 'CRITICAL';
  signal_count: number;
  dominant_signal: string;
  supporting_signals: string[];
  evidence_strength: 'LOW' | 'MODERATE' | 'STRONG' | 'VERY_STRONG';
  composite_risk_score: number;
  evidence_chain: string[];
  recommended_action: string;
  model_validation: AlertValidationContext;
  priority_score?: number;
  priority_rank?: number;
  priority_reason?: string;
}

export interface AlertSummary {
  total_alerts: number;
  critical_alerts_count: number;
  high_alerts_count: number;
  elevated_alerts_count: number;
  watch_alerts_count: number;
  info_alerts_count: number;
  states_under_watch_count: number;
  districts_under_watch_count: number;
}

export interface MonitoringOverview {
  active_alerts_count: number;
  high_critical_count: number;
  states_under_watch: number;
  districts_under_watch: number;
  persistent_signals_count: number;
  model_monitoring_status: string;
  latest_observation_year: number;
  summary: AlertSummary;
  scientific_disclaimer: string;
}

export interface StateWarningMapItem {
  state: string;
  state_code: number;
  severity: 'INFO' | 'WATCH' | 'ELEVATED' | 'HIGH' | 'CRITICAL';
  active_signals_count: number;
  critical_count: number;
  high_count: number;
  elevated_count: number;
  watch_count: number;
  average_yield_kg_ha: number;
  total_districts_monitored: number;
  dominant_concern: string;
}

export interface ChangeDetectionResult {
  metric_name: string;
  cusum_analysis: {
    change_detected: boolean;
    change_type: string;
    max_cusum_statistic: number;
    threshold: number;
    inflection_index: number | null;
    cusum_positive: number[];
    cusum_negative: number[];
    scientific_note: string;
  };
  trend_break_analysis: {
    trend_break_detected: boolean;
    inflection_year: number | null;
    pre_break_slope: number;
    post_break_slope: number;
    slope_delta: number;
    scientific_note: string;
  };
  volatility_shift_cv_delta: number;
  overall_change_flag: boolean;
  scientific_disclaimer: string;
  location?: string;
  state?: string;
  district?: string;
}

export interface WarningBacktest {
  total_evaluations: number;
  true_positives: number;
  false_positives: number;
  false_negatives: number;
  true_negatives: number;
  precision: number;
  recall: number;
  f1_score: number;
  false_positive_rate: number;
  false_negative_rate: number;
  alert_frequency_pct: number;
  mean_lead_time_years: number;
  evaluation_years_range: string;
  parameters: {
    yield_drop_threshold_pct: number;
    warning_zscore_threshold: number;
    lead_time_years: number;
  };
  is_chronologically_valid: boolean;
  scientific_disclaimer: string;
}

export interface MonitoringHealth {
  status: string;
  overall_health_score: number;
  data_quality_score: number;
  drift_status: string;
  prediction_mae: number;
  prediction_r2: number;
  calibration_quality: string;
  data_freshness_label: string;
  metrics_breakdown: {
    population_stability_index: number;
    data_completeness_pct: number;
    error_within_10pct_share: number;
    high_error_record_count: number;
    calibration_slope: number;
  };
  scientific_note: string;
}

export interface BacktestRequest {
  yield_drop_threshold_pct: number;
  warning_zscore_threshold: number;
  lead_time_years: number;
}

File: backend/schemas/monitoring.py
"""
Agricultural Monitoring & Early Warning Pydantic Schemas.

Defines validated data structures for temporal monitoring, early warning signals,
fused alerts, change detection, model health tracking, and historical backtesting.
"""

from __future__ import annotations

from typing import Dict, Any, List, Optional
from pydantic import BaseModel, Field


class TemporalSignal(BaseModel):
    metric_name: str
    record_count: int
    latest_year: Optional[int] = None
    latest_value: float
    yoy_change_pct: float
    yoy_change_absolute: float
    rolling_3yr_mean: float
    rolling_3yr_std: float
    rolling_3yr_zscore: float
    rolling_5yr_mean: float
    rolling_5yr_std: float
    rolling_5yr_zscore: float
    rolling_8yr_mean: float
    rolling_8yr_std: float
    trend_slope: float
    acceleration: float
    volatility_cv: float
    historical_mean: float
    deviation_from_historical: float
    trajectory: List[Dict[str, Any]] = []


class EarlyWarning(BaseModel):
    signal_id: str
    signal_type: str
    state: str
    district: str
    year: int
    severity: str
    trigger_value: float
    threshold: float
    unit: str
    evidence: List[str]
    recommended_action: str
    is_active: bool


class AlertValidationContext(BaseModel):
    r2: float
    mae: float
    drift_status: str
    data_quality_score: float


class Alert(BaseModel):
    alert_id: str
    location: str
    state: str
    district: str
    year: int
    severity: str
    signal_count: int
    dominant_signal: str
    supporting_signals: List[str]
    evidence_strength: str
    composite_risk_score: float
    evidence_chain: List[str]
    recommended_action: str
    model_validation: AlertValidationContext
    priority_score: Optional[float] = None
    priority_rank: Optional[int] = None
    priority_reason: Optional[str] = None


class AlertSummary(BaseModel):
    total_alerts: int
    critical_alerts_count: int
    high_alerts_count: int
    elevated_alerts_count: int
    watch_alerts_count: int
    info_alerts_count: int
    states_under_watch_count: int
    districts_under_watch_count: int


class MonitoringOverview(BaseModel):
    active_alerts_count: int
    high_critical_count: int
    states_under_watch: int
    districts_under_watch: int
    persistent_signals_count: int
    model_monitoring_status: str
    latest_observation_year: int
    summary: AlertSummary
    scientific_disclaimer: str


class RiskSignal(BaseModel):
    signal_type: str
    severity: str
    trigger_value: float
    threshold: float
    unit: str
    description: str


class ChangeDetectionResult(BaseModel):
    metric_name: str
    cusum_analysis: Dict[str, Any]
    trend_break_analysis: Dict[str, Any]
    volatility_shift_cv_delta: float
    overall_change_flag: bool
    scientific_disclaimer: str


class WarningBacktest(BaseModel):
    total_evaluations: int
    true_positives: int
    false_positives: int
    false_negatives: int
    true_negatives: int
    precision: float
    recall: float
    f1_score: float
    false_positive_rate: float
    false_negative_rate: float
    alert_frequency_pct: float
    mean_lead_time_years: int
    evaluation_years_range: str
    parameters: Dict[str, Any]
    is_chronologically_valid: bool
    scientific_disclaimer: str


class MonitoringHealth(BaseModel):
    status: str
    overall_health_score: float
    data_quality_score: float
    drift_status: str
    prediction_mae: float
    prediction_r2: float
    calibration_quality: str
    data_freshness_label: str
    metrics_breakdown: Dict[str, Any]
    scientific_note: str


class MonitoringHistory(BaseModel):
    total_records: int
    history: List[Alert]


class MonitoringQueryRequest(BaseModel):
    state: Optional[str] = Field(None, description="Filter by state")
    district: Optional[str] = Field(None, description="Filter by district")
    severity: Optional[str] = Field(None, description="Filter by minimum severity")
    signal_type: Optional[str] = Field(None, description="Filter by signal type")
    year: Optional[int] = Field(None, description="Filter by observation year")
    limit: int = Field(50, ge=1, le=200, description="Max alerts to return")


class BacktestRequest(BaseModel):
    yield_drop_threshold_pct: float = Field(-10.0, ge=-50.0, le=-1.0, description="Adverse threshold in %")
    warning_zscore_threshold: float = Field(1.2, ge=0.5, le=4.0, description="Warning trigger z-score")
    lead_time_years: int = Field(1, ge=1, le=3, description="Evaluation lead time in years")
